test/gurka/test_bidir_search.cc

```diff
@@ -0,0 +1,125 @@
+  auto map = gurka::buildtiles(layout, ways, {}, {}, tile_dir);
+  std::shared_ptr<baldr::GraphReader> reader =
+      test::make_clean_graphreader(map.config.get_child("mjolnir"));
+
+  // mark CD, DE & EF not_thru in both directions
+  std::vector<GraphId> not_thru_edgeids;
+  not_thru_edgeids.push_back(std::get<0>(gurka::findEdgeByNodes(*reader, layout, "C", "D")));
+  not_thru_edgeids.push_back(std::get<0>(gurka::findEdgeByNodes(*reader, layout, "D", "C")));
+  not_thru_edgeids.push_back(std::get<0>(gurka::findEdgeByNodes(*reader, layout, "D", "E")));
+  not_thru_edgeids.push_back(std::get<0>(gurka::findEdgeByNodes(*reader, layout, "E", "D")));
+  not_thru_edgeids.push_back(std::get<0>(gurka::findEdgeByNodes(*reader, layout, "E", "F")));
+  not_thru_edgeids.push_back(std::get<0>(gurka::findEdgeByNodes(*reader, layout, "F", "E")));
+
+  test::customize_edges(map.config, [&not_thru_edgeids](const GraphId& edgeid, DirectedEdge& edge) {
+    if (std::find(not_thru_edgeids.begin(), not_thru_edgeids.end(), edgeid) !=
+        not_thru_edgeids.end()) {
+      edge.set_not_thru(true);
+    }
+  });
+
+  auto result = gurka::do_action(valhalla::Options::route, map, {"A", "F"}, "auto", {});
+  gurka::assert::raw::expect_path(result, {"AB", "BC", "CD", "DE", "EF"});
+}
+
+TEST(StandAlone, failed_search) {
+  const std::string ascii_map = R"(
+  A-B---C---D---E-F
+  )";
+
+  // Make the 3 not_thru edges slow
+  const gurka::ways ways = {{"AB", {{"highway", "motorway"}}},
+                            {"BC", {{"highway", "secondary"}, {"maxspeed", "10"}}},
+                            {"CD", {{"highway", "secondary"}, {"maxspeed", "10"}}},
+                            {"DE", {{"highway", "secondary"}, {"maxspeed", "10"}}},
+                            {"EF", {{"highway", "motorway"}}}};
+  const auto layout = gurka::detail::map_to_coordinates(ascii_map, 100);
+  auto map = gurka::buildtiles(layout, ways, {}, {}, tile_dir);
+  std::shared_ptr<baldr::GraphReader> reader =
+      test::make_clean_graphreader(map.config.get_child("mjolnir"));
+
+  // mark CD, DE & EF not_thru in both directions
+  std::vector<GraphId> not_thru_edgeids;
+  not_thru_edgeids.push_back(std::get<0>(gurka::findEdgeByNodes(*reader, layout, "B", "C")));
+  not_thru_edgeids.push_back(std::get<0>(gurka::findEdgeByNodes(*reader, layout, "C", "B")));
+  not_thru_edgeids.push_back(std::get<0>(gurka::findEdgeByNodes(*reader, layout, "C", "D")));
+  not_thru_edgeids.push_back(std::get<0>(gurka::findEdgeByNodes(*reader, layout, "D", "C")));
+  not_thru_edgeids.push_back(std::get<0>(gurka::findEdgeByNodes(*reader, layout, "D", "E")));
+  not_thru_edgeids.push_back(std::get<0>(gurka::findEdgeByNodes(*reader, layout, "E", "D")));
+
+  test::customize_edges(map.config, [&not_thru_edgeids](const GraphId& edgeid, DirectedEdge& edge) {
+    if (std::find(not_thru_edgeids.begin(), not_thru_edgeids.end(), edgeid) !=
+        not_thru_edgeids.end()) {
+      edge.set_not_thru(true);
```

Comment:
its great to see you make use of that! i thought that one day opening that up for testing would be useful!

Reply:
I do not think this is a valid test case. I believe that not_thru will only be set in one direction. A directed edge that enters an area with one entrance/exit is marked as not_thru but the opposing edge that exits that area will not be marked as not_thru. Perhaps the not_thru logic is too aggressive that it somehow is marking some edges as not_thru in both directions, but the design is that only one direction of an edge should ever be marked as not_thru.

Reply:
i agree that its not true to what will actually get created in the data but its handy to be able to force the router into this situation even though its contrived. it would be better to show that it naturally happens for derived things like not_through, but at the same time the you have to figure out and spend tons of time designing the layout of the test to get it to happen.

Reply:
oh you have no idea how much time I spent on trying to recreate a scenario in gurka which would mark an edhe as `not_thru`! Thx to @genadz for pointing me to an existing test which was using this method.

BTW, `customize_edges` does not work if we're also setting live traffic conditions on the edge. When using live traffic, we pass-in a cleanly (I believe to avoid caching) created "reader" which does not have the customizations created in `customize_edges`. I think it has to do with live traffic & `customize_edges` altering different in-memory representations of the tile (I have to dig in further for the exact root-cause though).

Reply:
> I do not think this is a valid test case. I believe that not_thru will only be set in one direction. A directed edge that enters an area with one entrance/exit is marked as not_thru but the opposing edge that exits that area will not be marked as not_thru. Perhaps the not_thru logic is too aggressive that it somehow is marking some edges as not_thru in both directions, but the design is that only one direction of an edge should ever be marked as not_thru.

You're right, this is a contrived scenario with not thru's. However, such a scenario can happen when routing out of closures (edges marked closed due to live traffic conditions). In this case, the reverse search will not be able to expand into a closure.

I originally hit this condition while testing something with closures. ButI wanted to keep the test simple without adding live traffic into the mix, so resorted to making a test with not_thru's.